New plant heat pump features and supervisory controller

rraustad opened the pull request:
Pull request overview
---------------------
 - New Feature: Chiller plant support for advanced plant configurations
 - The use of heat pumps as plant equipment requires knowledge of plant loads to dispatch the heat pumps in the correct operating mode. A new supervisory controller will poll building and plant loads to dispatch plant equipment according to the user specified plant operation schemes.
 - This branch also updates how PlantLoop:HeatPump:EIR object's EIR as a function of PLR curves are interpreted. EIR curves should be proportional to part load ratio so example file curve objects were updated to reflect this concept.
 - There is also a minor correction to Coil Sizing table reports to not calculate coil design entering wet-bulb temperature when coil design entering humidity ratio is unknown (i.e., it will appear as -999.0 since the data is not available for calculation).
 - Closes #10133 that was used for testing and documenting differences found due to these changes.

**NOTE: ENHANCEMENTS MUST FOLLOW A SUBMISSION PROCESS INCLUDING A FEATURE PROPOSAL AND DESIGN DOCUMENT PRIOR TO SUBMITTING CODE**

### Pull Request Author
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [x] Title of PR should be user-synopsis style (clearly understandable in a standalone changelog context)
 - [x] Label the PR with at least one of: Defect, Refactoring, NewFeature, Performance, and/or DoNoPublish
 - [x] Pull requests that impact EnergyPlus code must also include unit tests to cover enhancement or defect repair
 - [x] Author should provide a "walkthrough" of relevant code changes using a GitHub code review comment process
 - [ ] If any diffs are expected, author must demonstrate they are justified using plots and descriptions
 - [ ] If changes fix a defect, the fix should be demonstrated in plots and descriptions
 - [ ] If any defect files are updated to a more recent version, upload new versions here or on DevSupport
 - [ ] If IDD requires transition, transition source, rules, ExpandObjects, and IDFs must be updated, and add IDDChange label
 - [ ] If structural output changes, add to output rules file and add OutputChange label
 - [ ] If adding/removing any LaTeX docs or figures, update that document's CMakeLists file dependencies

### To Do:
 - [ ] Convert new input files to hourly reporting and only for 2 select weather periods (or DDs only)

### Reviewer
This will not be exhaustively relevant to every PR.
 - [ ] Perform a Code Review on GitHub
 - [ ] If branch is behind develop, merge develop and build locally to check for side effects of the merge
 - [ ] If defect, verify by running develop branch and reproducing defect, then running PR and reproducing fix
 - [ ] If feature, test running new feature, try creative ways to break it
 - [ ] CI status: all green or justified
 - [ ] Check that performance is not impacted (CI Linux results include performance check)
 - [ ] Run Unit Test(s) locally
 - [ ] Check any new function arguments for performance impacts
 - [ ] Verify IDF naming conventions and styles, memos and notes and defaults
 - [ ] If new idf included, locally check the err file and other outputs

rraustad committed:
Move ChillerEIR code to E+

rraustad committed:
Initial NFP and Chiller Heater changes

rraustad committed:
Add new PlantEquipmentOperation:ChillerHeaterChangeover object

rraustad committed:
NFP - Chiller Plant Support for Carbon Initiatives

rraustad committed:
Update IORef docs

rraustad committed:
Merge branch 'develop' of https://github.com/NREL/EnergyPlus into NFP-Chiller-Plant-Support-for-Carbon-Initiatives

rraustad committed:
Update SetupOutputVariable

rraustad committed:
final change [decent_ci_skip]

rraustad committed:
update NFP [decent_ci_skip]

rraustad committed:
Remove endline comments [decent_ci_skip]

rraustad commented:
Some results from the new example file. The supervisory controller polls building loads and dispatches plant equipment. The following shows what the supervisory controller building loads profiles looks like. Updated based on most recent commit.

![image](https://user-images.githubusercontent.com/5778156/235743119-034f626f-6e95-4226-813b-88987321bcfd.png)

A couple zone temperatures to show zone control:

![image](https://user-images.githubusercontent.com/5778156/235743249-abeb6ecd-d477-4b6c-91e8-b8ebe1374f69.png)

rraustad committed:
Update NFP [decent_ci_skip] [actions skip]

rraustad committed:
Add unit test

rraustad committed:
[decent_ci_skip] [actions skip]

rraustad committed:
Minor update to supervisor inputs [decent_ci_skip] [actions skip]

rraustad committed:
Update recent changes to plant supervisor control [decent_ci_skip] [actions skip]

rraustad commented:
I am looking into this error and see that the air loops are cycling on and off. Since this NREL example file is a near copy of the Trane version I suspect this is happening there as well. Something is cycling the airloops. and plants?

    ** Severe  ** FindRootSimpleController: Controller error for controller = "VAV_3_COOLC_CONTROLLER"
    **   ~~~   **  Environment=ANNUAL, at Simulation time=01/16 13:15 - 13:30
    **   ~~~   **   Controller function is inconsistent with user specified controller action = Reverse action
    **   ~~~   **   Actuator will be set to maximum action
    **   ~~~   ** Controller control type=Temperature
    **   ~~~   ** Controller temperature setpoint = 7.88 [C]
    **   ~~~   ** Controller sensed temperature = 20.64 [C]
    **   ~~~   ** Controller actuator mass flow rate set to 12.59 [kg/s]
    **   ~~~   ** Controller actuator temperature = 20.68 [C]

Figure is day 1/16:

![image](https://user-images.githubusercontent.com/5778156/236065736-eb37cb40-b4db-4773-82d9-45da842b2f4a.png)

EnergyArchmage commented:
That's night cycle availability manager kicking on air systems for fixed periods of time.

Get Outlook for Android<https://aka.ms/AAb9ysg>

________________________________
From: Richard Raustad ***@***.***>
Sent: Wednesday, May 3, 2023, 6:37 PM
To: NREL/EnergyPlus ***@***.***>
Cc: brent.griffith energyarchmage.com ***@***.***>; Assign ***@***.***>
Subject: Re: [NREL/EnergyPlus] NFP-Chiller Plant Support for Carbon Initiatives (PR #9979)


I am looking into this error and see that the air loops are cycling on and off. Since this NREL example file is a near copy of the Trane version I suspect this is happening there as well. Something is cycling the airloops. and plants?

** Severe  ** FindRootSimpleController: Controller error for controller = "VAV_3_COOLC_CONTROLLER"
**   ~~~   **  Environment=ANNUAL, at Simulation time=01/16 13:15 - 13:30
**   ~~~   **   Controller function is inconsistent with user specified controller action = Reverse action
**   ~~~   **   Actuator will be set to maximum action
**   ~~~   ** Controller control type=Temperature
**   ~~~   ** Controller temperature setpoint = 7.88 [C]
**   ~~~   ** Controller sensed temperature = 20.64 [C]
**   ~~~   ** Controller actuator mass flow rate set to 12.59 [kg/s]
**   ~~~   ** Controller actuator temperature = 20.68 [C]


Figure is day 1/16:

[image]<https://user-images.githubusercontent.com/5778156/236065736-eb37cb40-b4db-4773-82d9-45da842b2f4a.png>

—
Reply to this email directly, view it on GitHub<https://github.com/NREL/EnergyPlus/pull/9979#issuecomment-1533840628>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/ACCZMUM5JLGRUNRZYTURXC3XELM27ANCNFSM6AAAAAAXKFPQCA>.
You are receiving this because you were assigned.Message ID: ***@***.***>

________________________________
From: Richard Raustad ***@***.***>
Sent: Wednesday, May 3, 2023 6:37:35 PM
To: NREL/EnergyPlus ***@***.***>
Cc: brent.griffith energyarchmage.com ***@***.***>; Assign ***@***.***>
Subject: Re: [NREL/EnergyPlus] NFP-Chiller Plant Support for Carbon Initiatives (PR #9979)


I am looking into this error and see that the air loops are cycling on and off. Since this NREL example file is a near copy of the Trane version I suspect this is happening there as well. Something is cycling the airloops. and plants?

** Severe  ** FindRootSimpleController: Controller error for controller = "VAV_3_COOLC_CONTROLLER"
**   ~~~   **  Environment=ANNUAL, at Simulation time=01/16 13:15 - 13:30
**   ~~~   **   Controller function is inconsistent with user specified controller action = Reverse action
**   ~~~   **   Actuator will be set to maximum action
**   ~~~   ** Controller control type=Temperature
**   ~~~   ** Controller temperature setpoint = 7.88 [C]
**   ~~~   ** Controller sensed temperature = 20.64 [C]
**   ~~~   ** Controller actuator mass flow rate set to 12.59 [kg/s]
**   ~~~   ** Controller actuator temperature = 20.68 [C]


Figure is day 1/16:

[image]<https://user-images.githubusercontent.com/5778156/236065736-eb37cb40-b4db-4773-82d9-45da842b2f4a.png>

—
Reply to this email directly, view it on GitHub<https://github.com/NREL/EnergyPlus/pull/9979#issuecomment-1533840628>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/ACCZMUM5JLGRUNRZYTURXC3XELM27ANCNFSM6AAAAAAXKFPQCA>.
You are receiving this because you were assigned.Message ID: ***@***.***>

EnergyArchmage committed:
refactor supervisory control break up long routine into pieces, use current plant loads in equipment dispatch

add and revise output variables from supervisory operation scheme.

revert draft changes to plant component temperature sources as out of scope for this project, was for early testing of supervisory routines

rraustad committed:
Rework EIR sizing for companion coils sizing method

rraustad committed:
Tentative correction for 9991

rraustad committed:
Merge branch 'develop' of https://github.com/NREL/EnergyPlus into NFP-Chiller-Plant-Support-for-Carbon-Initiatives

rraustad committed:
[decent_ci_skip] [actions skip]

rraustad committed:
Merge branch 'NFP-Chiller-Plant-Support-for-Carbon-Initiatives' of https://github.com/NREL/EnergyPlus into NFP-Chiller-Plant-Support-for-Carbon-Initiatives

rraustad committed:
small change

rraustad committed:
[decent_ci_skip] [actions skip]

rraustad committed:
clean up actions results

rraustad committed:
Clean up unit tests and EIR sizing

EnergyArchmage committed:
draft out I O Ref content for supervisory plant op scheme [decent_ci_skip] [actions skip]

rraustad committed:
Merge branch 'NFP-Chiller-Plant-Support-for-Carbon-Initiatives' of https://github.com/NREL/EnergyPlus into NFP-Chiller-Plant-Support-for-Carbon-Initiatives

rraustad committed:
Merge branch 'develop' of https://github.com/NREL/EnergyPlus into NFP-Chiller-Plant-Support-for-Carbon-Initiatives

rraustad committed:
clang format PlantHeatPumpEIR unit test

rraustad committed:
Clean up warnings and tweak sizing of heat recovery water flow rate

rraustad committed:
Merge branch 'develop' of https://github.com/NREL/EnergyPlus into NFP-Chiller-Plant-Support-for-Carbon-Initiatives

rraustad committed:
Resolve sizing method for EIR chillers and a few other issues

EnergyArchmage committed:
add test file for 2 air source heat pumps serving four pipe beam large office

EnergyArchmage committed:
revise output variable names and description [decent_ci_skip] [actions skip]

change names as suggested, revise inputs and I/O ref description

fix EIR func of PLR curve in new testfile

rraustad committed:
A little clean up

rraustad committed:
Merge branch 'NFP-Chiller-Plant-Support-for-Carbon-Initiatives' of https://github.com/NREL/EnergyPlus into NFP-Chiller-Plant-Support-for-Carbon-Initiatives

rraustad committed:
another pass on code and docs

rraustad committed:
Update unit test based on latest code

rraustad committed:
Merge branch 'develop' of https://github.com/NREL/EnergyPlus into NFP-Chiller-Plant-Support-for-Carbon-Initiatives

EnergyArchmage committed:
Supervisory control, boiler controls [decent_CI_Skip] [actions no]

add logic to treat auxiliary and supplemental boilers differently,

IO ref  updates,

change testfile to all electric

EnergyArchmage committed:
call supervisory control every iteration

improves control of plant temperatures

EnergyArchmage committed:
supervisory control IO ref edits

EnergyArchmage commented on doc/input-output-reference/src/overview/group-plant-equipment.tex:
```diff
@@ -5319,7 +5319,7 @@ \subsubsection{Outputs}\label{outputs-13-003}
 
 \subsection{HeatPump:PlantLoop:EIR:Cooling (and Heating)}\label{plhp_eir}
 
```
The Cooling and Heating versions are diverging inputs with these changes, for example heating version has defrost stuff.  I think this I/O reference entry will need to be split up into two in order to handle the input descriptions. cleanly.  

And more description of new air-source capababilities.

EnergyArchmage commented on src/EnergyPlus/PlantLoopHeatPumpEIR.cc:
```diff
@@ -361,14 +569,30 @@ void EIRPlantLoopHeatPump::doPhysics(EnergyPlusData &state, Real64 currentLoad)
+    this->oneTimeInit(state);          // plant setup
+    this->isPlantInletOrOutlet(state); // check location
 
     if (state.dataGlobal->BeginEnvrnFlag && this->envrnInit && state.dataPlnt->PlantFirstSizesOkayToFinalize) {
```
Though not really part of this PR, find that this routine is not coded properly for iterative plant sizing and HVAC Sizing Simulations.  Plant components should continually resize during a set of plant sizing iterations and then report when finalized.  Putting the sizing calls behind this if statement is not going to work properly because it would not size during early iterations (PlantFirstSizesOkayToFinalize = false) nor will it reenter multiple times once this->envrnInit gets set to false.  

Testing with coincident plant sizing using HVAC Sizing Simulations showed that final sizes are not reported, only initial sizes.

EnergyArchmage committed:
Supervisory plant control and plant loop heat pump EIR model

 revise how EIR heat pump sizing is called to ensure iterative sizing and support HVAC Sizing Simulations

supervisory chilled water plant loop loads based on outlet of chilled water buffer tank.

rraustad committed:
allow multiple sizing passes and use source Cp

rraustad committed:
better loop exit strategy

rraustad committed:
update unit tests

rraustad committed:
Merge branch 'develop' of https://github.com/NREL/EnergyPlus into NFP-Chiller-Plant-Support-for-Carbon-Initiatives

rraustad committed:
Merge branch 'NFP-Chiller-Plant-Support-for-Carbon-Initiatives' of https://github.com/NREL/EnergyPlus into NFP-Chiller-Plant-Support-for-Carbon-Initiatives

rraustad committed:
Update getInput to allow a single plant HP

rraustad committed:
Merge branch 'develop' of https://github.com/NREL/EnergyPlus into NFP-Chiller-Plant-Support-for-Carbon-Initiatives

rraustad committed:
Merge branch 'develop' of https://github.com/NREL/EnergyPlus into NFP-Chiller-Plant-Support-for-Carbon-Initiatives

rraustad committed:
IO Ref and getInput update

rraustad committed:
Merge branch 'develop' of https://github.com/NREL/EnergyPlus into NFP-Chiller-Plant-Support-for-Carbon-Initiatives

rraustad committed:
Eng Ref update, correct dry coil fraction calculation, update idd.

rraustad committed:
Merge branch 'develop' of https://github.com/NREL/EnergyPlus into NFP-Chiller-Plant-Support-for-Carbon-Initiatives

rraustad committed:
Oops

rraustad commented on src/EnergyPlus/PlantLoopHeatPumpEIR.cc:
```diff
@@ -254,9 +328,67 @@ void EIRPlantLoopHeatPump::doPhysics(EnergyPlusData &state, Real64 currentLoad)
+                availableCapacity *= dryCorrectionFactor;
+            } else {
+                // interpolation of heating capacity between wet and dry is based on outdoor relative humidity over 60%-90% range
+                Real64 semiDryFactor = dryCorrectionFactor + (1.0 - dryCorrectionFactor) * (1.0 - ((RH90 - state.dataEnvrn->OutRelHum) / rangeRH));
```
@EnergyArchmage this equation was incorrect. I added the `1.0 -` in `(1.0 - ((RH90 -` to get this math right.

rraustad committed:
Clean up doc text

rraustad committed:
Add secondary plant temperature limits and use plant loads in dispatch

rraustad committed:
Pull recent plant supervisory changes

rraustad committed:
Merge branch 'develop' of https://github.com/NREL/EnergyPlus into NFP-Chiller-Plant-Support-for-Carbon-Initiatives

rraustad committed:
Correct typo

rraustad committed:
[decent_ci_skip] [actions skip]

rraustad committed:
Merge branch 'develop' of https://github.com/NREL/EnergyPlus into NFP-Chiller-Plant-Support-for-Carbon-Initiatives

rraustad committed:
Update new model with recent changes

rraustad committed:
Merge branch 'develop' of https://github.com/NREL/EnergyPlus into NFP-Chiller-Plant-Support-for-Carbon-Initiatives

rraustad commented on src/EnergyPlus/PlantLoopHeatPumpEIR.cc:
```diff
@@ -188,12 +210,38 @@ void EIRPlantLoopHeatPump::setOperatingFlowRatesWSHP(EnergyPlusData &state)
+                (this->loadSideIsPlantInlet) ? state.dataLoopNodes->Node(this->loadSideNodes.inlet).MassFlowRate : this->loadSideDesignMassFlowRate;
+            this->sourceSideMassFlowRate = (this->sourceSideIsPlantOutlet) ? state.dataLoopNodes->Node(this->sourceSideNodes.inlet).MassFlowRate
+                                                                           : this->sourceSideDesignMassFlowRate;
+
```
The above was meant to identify the HR HP. The `= (this->sourceSideIsPlantOutlet) ?` is shutting down the chiller where loadSideIsPlantInlet = true and sourceSideIsPlantOutput = false. I changed this to:

            this->loadSideMassFlowRate = (this->loadSideIsPlantInlet && this->sourceSideIsPlantOutlet)
                                             ? state.dataLoopNodes->Node(this->loadSideNodes.inlet).MassFlowRate
                                             : this->loadSideDesignMassFlowRate;
            this->sourceSideMassFlowRate = (this->loadSideIsPlantInlet && this->sourceSideIsPlantOutlet)
                                               ? state.dataLoopNodes->Node(this->sourceSideNodes.inlet).MassFlowRate
                                               : this->sourceSideDesignMassFlowRate;

and now the example file PlantLoopHeatPump_EIR_WaterSource runs again. I should change this to a single bool that is true if loadSideIsPlantInlet = true and sourceSideIsPlantOutput = true.

rraustad committed:
Clean up logic for heat recovery heat pump

rraustad committed:
Merge branch 'develop' of https://github.com/NREL/EnergyPlus into NFP-Chiller-Plant-Support-for-Carbon-Initiatives

rraustad committed:
Address warning and execution speed

rraustad commented on src/EnergyPlus/PlantUtilities.cc:
```diff
@@ -1023,18 +1024,23 @@ void UpdateChillerComponentCondenserSide(EnergyPlusData &state,
+                OtherLoopNum = state.dataPlnt->PlantLoop(LoopNum).LoopSide(LoopSide).Connected(ConnectLoopNum).LoopNum;
+                OtherLoopSide = state.dataPlnt->PlantLoop(LoopNum).LoopSide(LoopSide).Connected(ConnectLoopNum).LoopSideNum;
+                state.dataPlnt->PlantLoop(OtherLoopNum).LoopSide(OtherLoopSide).SimLoopSideNeeded = true;
+                // }
```
These 2 changes, 1 here and 1 below, are causing diffs from the additional plant simulations. I think I'll create a branch with these specific PlantUtility changes to verify and possibly merge before this branch. The only diffs I have seen up until the lastest commits were in the PlantLoopHeatPump_EIR*.idf files (very small diffs). Now there are many larger diffs because of these couple changes.

rraustad committed:
Revise defaults to be more reasonable

rraustad committed:
Merge branch 'develop' of https://github.com/NREL/EnergyPlus into NFP-Chiller-Plant-Support-for-Carbon-Initiatives

rraustad committed:
Ok clang, I see that.

EnergyArchmage committed:
remove first hvac iteration logic from heat pump supervisory control

found that during plant iterations machines can get turned off after being turned on so need to always init off.

rraustad committed:
Merge branch 'NFP-Chiller-Plant-Support-for-Carbon-Initiatives' of https://github.com/NREL/EnergyPlus into NFP-Chiller-Plant-Support-for-Carbon-Initiatives

EnergyArchmage committed:
Clean up testfile

add output variables for secondary plant loads

rraustad committed:
Merge branch 'NFP-Chiller-Plant-Support-for-Carbon-Initiatives' of https://github.com/NREL/EnergyPlus into NFP-Chiller-Plant-Support-for-Carbon-Initiatives

rraustad committed:
Avoid unnecessary calls to same supervisor managing multiple plants

rraustad committed:
Correct coding mistake

rraustad committed:
Correct unit tests to fix CI build issue

rraustad committed:
Clean up unused code and latex build warnings

rraustad committed:
Revise comment

rraustad commented on src/EnergyPlus/Plant/EquipAndOperations.cc:
```diff
@@ -0,0 +1,1718 @@
+// EnergyPlus, Copyright (c) 1996-2023, The Board of Trustees of the University of Illinois,
```
New source that holds all the supervisor logic

rraustad commented on src/EnergyPlus/Plant/EquipAndOperations.cc:
```diff
@@ -0,0 +1,1718 @@
+        InitAirSourcePlantEquipmentOff(state);
+        ProcessAndSetAirSourcePlantEquipLists(state);
+        ProcessAndSetDedicatedHeatRecovWWHP(state);
+        ProcessAndSetAuxilBoiler(state);
```
Dedicated worker functions to find building loads, plant loads, determine dispatch logic for equipment, initialize plant equipment, dispatch plant equipment, dispatch heat recovery equipment, and dispatch boiler if needed.

rraustad commented on src/EnergyPlus/Plant/EquipAndOperations.cc:
```diff
@@ -0,0 +1,1718 @@
+
+    void ChillerHeaterSupervisoryOperationData::DetermineCurrentBuildingLoads(EnergyPlusData &state)
+    {
+        // Poll the loads on the zones to help decide how to run
```
sum building loads

rraustad commented on src/EnergyPlus/Plant/EquipAndOperations.cc:
```diff
@@ -0,0 +1,1718 @@
+
+    void ChillerHeaterSupervisoryOperationData::DetermineCurrentPlantLoads(EnergyPlusData &state)
+    {
+
```
sum plant loads

rraustad commented on src/EnergyPlus/Plant/EquipAndOperations.cc:
```diff
@@ -0,0 +1,1718 @@
+        this->PlantOps.AirSourcePlantHeatingOnly = false;
+        this->PlantOps.AirSourcePlantSimultaneousHeatingAndCooling = false;
+        this->PlantOps.SimultaneousHeatingCoolingWithCoolingDominant = false;
+        this->PlantOps.SimultaneousHeatingCoolingWithHeatingDominant = false;
```
Set these control flags based on building and/or plant loads

rraustad commented on src/EnergyPlus/Plant/EquipAndOperations.cc:
```diff
@@ -0,0 +1,1718 @@
+    }
+
+    void ChillerHeaterSupervisoryOperationData::InitAirSourcePlantEquipmentOff(EnergyPlusData &state)
+    {
```
initialize plant equipment off

rraustad commented on src/EnergyPlus/Plant/EquipAndOperations.cc:
```diff
@@ -0,0 +1,1718 @@
+    }
+
+    void ChillerHeaterSupervisoryOperationData::ProcessAndSetAirSourcePlantEquipLists(EnergyPlusData &state)
+    {
```
enable plant equipment based on logical flags and loads

rraustad commented on src/EnergyPlus/Plant/EquipAndOperations.cc:
```diff
@@ -0,0 +1,1718 @@
+    }
+
+    void ChillerHeaterSupervisoryOperationData::ProcessAndSetDedicatedHeatRecovWWHP(EnergyPlusData &state)
+    {
```
determine if dedicated heat recovery heat pump should operate

rraustad commented on src/EnergyPlus/Plant/EquipAndOperations.cc:
```diff
@@ -0,0 +1,1718 @@
+    }
+
+    void ChillerHeaterSupervisoryOperationData::ProcessAndSetAuxilBoiler(EnergyPlusData &state)
+    {
```
Dispatch boiler if HPs are off

rraustad commented on src/EnergyPlus/Plant/EquipAndOperations.cc:
```diff
@@ -0,0 +1,1718 @@
+    }
+
+    Real64 ChillerHeaterSupervisoryOperationData::DetermineHWSetpointOARest(EnergyPlusData &state)
+    {
```
worker function to calculate plant set point temperature

rraustad commented on idd/Energy+.idd.in:
```diff
@@ -74711,18 +74705,62 @@ HeatPump:PlantLoop:EIR:Cooling,
         \type object-list
         \object-list UniVariateFunctions
         \note quadratic curve = a + b*PLR + c*PLR**2 is typical, other univariate curves may be used
         \note PLR = part load ratio (cooling load/steady state capacity)
```
add new fields for cooling EIR HP

rraustad commented on idd/Energy+.idd.in:
```diff
@@ -74810,14 +74842,125 @@ HeatPump:PlantLoop:EIR:Heating,
         \type object-list
         \object-list UnivariateFunctions
         \note quadratic curve = a + b*PLR + c*PLR**2 is typical, other univariate curves may be used
         \note PLR = part load ratio (hot load/steady state capacity)
```
add new fields for heating EIR HP

rraustad commented on idd/Energy+.idd.in:
```diff
@@ -83300,6 +83443,89 @@ PlantEquipmentOperation:OutdoorDewpointDifference,
        \type object-list
        \object-list PlantAndCondenserEquipmentLists
 
```
new supervisory controller

rraustad commented on src/EnergyPlus/Plant/Enums.hh:
```diff
@@ -241,10 +242,12 @@ enum class FlowMode
     NotModulated,
     LeavingSetpointModulated,
+    VariableSpeedPump,
     Num
```
allow variable speed pumps

rraustad commented on src/EnergyPlus/PlantCondLoopOperation.cc:
```diff
@@ -1833,6 +1847,557 @@ void FindCompSPInput(EnergyPlusData &state,
+{
+    // process input objects for advanced operation scheme, use json type input patterns
+
+    if (NumSchemes > 0 && state.dataPlantCondLoopOp->LoadSupervisoryChillerHeaterOpScheme) {
```
load schemes for supervisory controller

rraustad commented on src/EnergyPlus/PlantCondLoopOperation.cc:
```diff
@@ -2114,8 +2679,28 @@ void InitLoadDistribution(EnergyPlusData &state, bool const FirstHVACIteration)
                     } // Equipment on List
                 }     // List
-            }         // operation scheme
-        }             // loop
```
get pointer for supervisory controller

rraustad commented on src/EnergyPlus/PlantCondLoopOperation.cc:
```diff
@@ -2144,7 +2729,7 @@ void InitLoadDistribution(EnergyPlusData &state, bool const FirstHVACIteration)
                             dummy_op_scheme_1.EquipList(1).ListPtr = ListNum;
                             dummy_op_scheme_1.EquipList(1).CompPtr = EquipNum;
-                        } else if (dummy_loop_equip.NumOpSchemes > 0) { // already an op scheme
+                        } else { // already an op scheme
```
this is the else from if (NumOpSchemes == 0) so no need to check > 0

rraustad commented on src/EnergyPlus/PlantCondLoopOperation.cc:
```diff
@@ -2285,12 +2870,26 @@ void InitLoadDistribution(EnergyPlusData &state, bool const FirstHVACIteration)
+        // set sim flag so each supervisor is only simulated once in the plant loop below
+        for (DataPlant::ChillerHeaterSupervisoryOperationData &supervisor : state.dataPlantCondLoopOp->ChillerHeaterSupervisoryOperationSchemes) {
+            supervisor.needsSimulation = true;
+        }
```
logic to call supervisor only once in loop below

rraustad commented on src/EnergyPlus/PlantCondLoopOperation.cc:
```diff
@@ -2285,12 +2870,26 @@ void InitLoadDistribution(EnergyPlusData &state, bool const FirstHVACIteration)
+        for (DataPlant::ChillerHeaterSupervisoryOperationData &supervisor : state.dataPlantCondLoopOp->ChillerHeaterSupervisoryOperationSchemes) {
+            supervisor.needsSimulation = true;
+        }
         // Update the OpScheme schedules
```
each of these loops will call the same supervisor that only needs to be called once, otherwise the call to the supervisor would do the same thing for all the plants

rraustad commented on src/EnergyPlus/PlantCondLoopOperation.cc:
```diff
@@ -2338,6 +2937,27 @@ void InitLoadDistribution(EnergyPlusData &state, bool const FirstHVACIteration)
             //      !'call warning 'no current control scheme specified.  Loop Equipment will be shut down'
             //    ENDIF
         }
```
supervisor is currently called each iteration to converge on final plant operation

rraustad commented on src/EnergyPlus/PlantLoopHeatPumpEIR.cc:
```diff
@@ -92,7 +92,7 @@ void EIRPlantLoopHeatPump::simulate(
 
     if (this->waterSource) {
-        this->setOperatingFlowRatesWSHP(state);
+        this->setOperatingFlowRatesWSHP(state, FirstHVACIteration);
```
water flow rate initialization will act different on FirstHVACIteration, mostly for VS pumps

rraustad commented on src/EnergyPlus/PlantLoopHeatPumpEIR.cc:
```diff
@@ -108,17 +108,32 @@ void EIRPlantLoopHeatPump::simulate(
     }
 
     if (this->running) {
-        this->doPhysics(state, CurLoad);
```
For set point control calculate set point and HP load prior to calling doPhysics

rraustad commented on src/EnergyPlus/PlantLoopHeatPumpEIR.cc:
```diff
@@ -108,17 +108,32 @@ void EIRPlantLoopHeatPump::simulate(
         this->resetReportingVariables();
     }
 
     // update nodes
```
added some node safe copy to water nodes

rraustad commented on src/EnergyPlus/PlantLoopHeatPumpEIR.cc:
```diff
@@ -188,12 +210,37 @@ void EIRPlantLoopHeatPump::setOperatingFlowRatesWSHP(EnergyPlusData &state)
-            state, this->loadSideMassFlowRate, this->loadSideNodes.inlet, this->loadSideNodes.outlet, this->loadSidePlantLoc);
-        PlantUtilities::SetComponentFlowRate(
-            state, this->sourceSideMassFlowRate, this->sourceSideNodes.inlet, this->sourceSideNodes.outlet, this->sourceSidePlantLoc);
+        // apply min/max operating limits based on source side entering fluid temperature
```
heat recovey HPs are connected to plant inlet or outlet nodes so must pass thru water mass flow even when off

rraustad commented on src/EnergyPlus/PlantLoopHeatPumpEIR.cc:
```diff
@@ -188,12 +210,37 @@ void EIRPlantLoopHeatPump::setOperatingFlowRatesWSHP(EnergyPlusData &state)
-        PlantUtilities::SetComponentFlowRate(
-            state, this->sourceSideMassFlowRate, this->sourceSideNodes.inlet, this->sourceSideNodes.outlet, this->sourceSidePlantLoc);
+        // apply min/max operating limits based on source side entering fluid temperature
+        if (this->minSourceTempLimit > this->sourceSideInletTemp || this->maxSourceTempLimit < this->sourceSideInletTemp) {
```
heat recovery HP is off if temperature limits are exceeded

rraustad commented on src/EnergyPlus/PlantLoopHeatPumpEIR.cc:
```diff
@@ -188,12 +210,37 @@ void EIRPlantLoopHeatPump::setOperatingFlowRatesWSHP(EnergyPlusData &state)
+            this->loadSideMassFlowRate = (this->heatRecoveryHeatPump) ? state.dataLoopNodes->Node(this->loadSideNodes.inlet).MassFlowRate : 0.0;
+            this->sourceSideMassFlowRate = (this->heatRecoveryHeatPump) ? state.dataLoopNodes->Node(this->sourceSideNodes.inlet).MassFlowRate : 0.0;
+            this->running = false;
+        } else {
```
else the HP must operate

rraustad commented on src/EnergyPlus/PlantLoopHeatPumpEIR.cc:
```diff
@@ -188,12 +210,37 @@ void EIRPlantLoopHeatPump::setOperatingFlowRatesWSHP(EnergyPlusData &state)
+                (this->heatRecoveryHeatPump) ? state.dataLoopNodes->Node(this->loadSideNodes.inlet).MassFlowRate : this->loadSideDesignMassFlowRate;
+            this->sourceSideMassFlowRate = (this->heatRecoveryHeatPump) ? state.dataLoopNodes->Node(this->sourceSideNodes.inlet).MassFlowRate
+                                                                        : this->sourceSideDesignMassFlowRate;
+
```
HP mass flow logic based on FirstHVACIteration

rraustad commented on src/EnergyPlus/PlantLoopHeatPumpEIR.cc:
```diff
@@ -188,12 +210,37 @@ void EIRPlantLoopHeatPump::setOperatingFlowRatesWSHP(EnergyPlusData &state)
+                    }
+                }
+            }
+
```
let worker function set node flow rates

rraustad commented on src/EnergyPlus/PlantLoopHeatPumpEIR.cc:
```diff
@@ -254,9 +323,81 @@ void EIRPlantLoopHeatPump::doPhysics(EnergyPlusData &state, Real64 currentLoad)
 
     // evaluate capacity modifier curve and determine load side heat transfer
-    Real64 capacityModifierFuncTemp = Curve::CurveValue(state, this->capFuncTempCurveIndex, loadSideOutletSetpointTemp, this->sourceSideInletTemp);
+    // any adjustment to outlet water temp set point requires some form of iteration
```
next loop converges on operating capacity

rraustad commented on src/EnergyPlus/PlantLoopHeatPumpEIR.cc:
```diff
@@ -254,9 +323,81 @@ void EIRPlantLoopHeatPump::doPhysics(EnergyPlusData &state, Real64 currentLoad)
+
+        availableCapacity = this->referenceCapacity * capacityModifierFuncTemp;
+
+        // apply heating mode dry outdoor (evaporator) coil correction factor for air-cooled equipment
```
next is Trane dry condenser coil capacity correction model

rraustad commented on src/EnergyPlus/PlantLoopHeatPumpEIR.cc:
```diff
@@ -254,9 +323,81 @@ void EIRPlantLoopHeatPump::doPhysics(EnergyPlusData &state, Real64 currentLoad)
+        if (availableCapacity > 0) {
+            partLoadRatio = std::clamp(std::abs(currentLoad) / availableCapacity, 0.0, 1.0);
+        }
+
```
next is limits on min/max water temp and adjusting set point to maintain operation within limits

rraustad commented on src/EnergyPlus/PlantLoopHeatPumpEIR.cc:
```diff
@@ -254,9 +323,81 @@ void EIRPlantLoopHeatPump::doPhysics(EnergyPlusData &state, Real64 currentLoad)
+                loadSideOutletSetpointTemp = maxWaterTemp + (1.0 - partLoadRatio) * (originalLoadSideOutletSPTemp - maxWaterTemp);
+                waterTempExceeded = true;
+            }
+        }
```
next is checking that heat recovery HP operation will not overheat source side plant, if so then reduce PLR. Otherwise heat recovery HP can increase/decrease source side plant way out of bounds (i.e., significantly overheat or overcool)

rraustad commented on src/EnergyPlus/PlantLoopHeatPumpEIR.cc:
```diff
@@ -254,9 +323,81 @@ void EIRPlantLoopHeatPump::doPhysics(EnergyPlusData &state, Real64 currentLoad)
+            break;
+        }
+    }
+
```
next is HP minimum PLR code where HP cycles below min PLR (similar to other chillers and VRF)

rraustad commented on src/EnergyPlus/PlantLoopHeatPumpEIR.cc:
```diff
@@ -279,11 +420,85 @@ void EIRPlantLoopHeatPump::doPhysics(EnergyPlusData &state, Real64 currentLoad)
-    Real64 partLoadRatio = 0.0;
-    if (availableCapacity > 0) {
-        partLoadRatio = max(0.0, min(std::abs(currentLoad) / availableCapacity, 1.0));
+    // Initializing defrost adjustment factors
```
next is new defrost models

rraustad commented on src/EnergyPlus/PlantLoopHeatPumpEIR.cc:
```diff
@@ -279,11 +420,85 @@ void EIRPlantLoopHeatPump::doPhysics(EnergyPlusData &state, Real64 currentLoad)
+        this->defrostEnergyRate = 0.0;
+        this->loadDueToDefrost = 0.0;
+        this->fractionalDefrostTime = 0.0;
     }
```
and now we know operating available capacity and can store/report defrost energy use

rraustad commented on src/EnergyPlus/PlantLoopHeatPumpEIR.cc:
```diff
@@ -341,44 +559,60 @@ void EIRPlantLoopHeatPump::doPhysics(EnergyPlusData &state, Real64 currentLoad)
 
-    this->powerUsage = (this->loadSideHeatTransfer / this->referenceCOP) * eirModifierFuncPLR * eirModifierFuncTemp;
+    this->powerUsage =
+        (availableCapacity / this->referenceCOP) * eirModifierFuncPLR * eirModifierFuncTemp * InputPowerMultiplier * this->cyclingRatio;
```
the above was corrected to include cycling ratio (= 1 above min PLR)

rraustad replied:
@mjwitte here is the change that accompnies the change to example file EIRfPLR curves. The old equation gave full load power because eirModifierFuncPLR = 1. Now power is proportional to PLR and adds in a factor where below minPLR the HP is cycling on and off (cyclingRatio = 1 above minPLR). So I guess I didn't really change the equation, but corrected the input file curve.

rraustad replied:
OK, I remember this now, the old equation used this->loadSideHeatTransfer, instead of availableCapacity, which did account for PLR. So if someone used a proper EIRfPLR curve the power would double count PLR. The only transition I can think of is updating a unity curve.

mjwitte replied:
This adds to the age-old confusion of what an EIRfPLR curve should look like. If efficiency is flat vs plr should the curve be 1.0 constant (no modification) or should it be a 45-degree line? If you use `this->loadSideHeatTransfer` which I find more intuitive, then it should be a 1.0 constant curve for flat efficiency. What does the EIR chiller use?

rraustad replied:
The EIR chiller uses:

    this->Power = (AvailChillerCap / ReferenceCOP) * this->ChillerEIRFPLR * this->ChillerEIRFT * FRAC;

mjwitte replied:
Maybe I'm the only one who's old-age confused. The recently added HeatPump:AirToWater:FuelFired:Heating was based on this heat pump, so it's using the other method:
`2224: this->fuelRate = this->loadSideHeatTransfer / (this->referenceCOP * CRF) * eirModifierFuncPLR * eirModifierFuncTemp * eirDefrost;`
@jcyuan2020 Noting this here. Not sure if we need to change HeatPump:AirToWater:FuelFired:Heating or just make sure documentation is very clear and possibly add some checks/warning to catch eirfplr curves that are assuming the wrong basis.

jcyuan2020 replied:
@mjwitte Noted. 

This would definitely make some impacts the way how the heating/cooling energy is calculated for the plant-loop heat pumps or chillers---and it is likely that the users would find this difference in the results between the old and new versions soon. A previous discussion about this issue can be also found in PR #9792 (addressing earlier issue post #9252) about the differences of the two methods.

rraustad commented on src/EnergyPlus/PlantLoopHeatPumpEIR.cc:
```diff
@@ -341,44 +559,60 @@ void EIRPlantLoopHeatPump::doPhysics(EnergyPlusData &state, Real64 currentLoad)
-    } else if (this->airSource) {
+        auto &thisSourcePlantLoop = state.dataPlnt->PlantLoop(this->sourceSidePlantLoc.loopNum);
+        CpSrc = FluidProperties::GetSpecificHeatGlycol(
+            state, thisSourcePlantLoop.FluidName, this->sourceSideInletTemp, thisSourcePlantLoop.FluidIndex, "PLHPEIR::simulate()");
```
use source side Cp just like in #10048

rraustad commented on src/EnergyPlus/PlantLoopHeatPumpEIR.cc:
```diff
@@ -341,44 +559,60 @@ void EIRPlantLoopHeatPump::doPhysics(EnergyPlusData &state, Real64 currentLoad)
     }
     Real64 const sourceMCp = this->sourceSideMassFlowRate * CpSrc;
     this->sourceSideOutletTemp = this->calcSourceOutletTemp(this->sourceSideInletTemp, this->sourceSideHeatTransfer / sourceMCp);
+
```
this next code was to prevent souce side temps from getting out of control. code above replaces this but this was not removed in case that other code wasn't working correctly. We saw some pretty out of control temps when dispatching the heat recovery HP, on the source side, because load side was controlling and source side just put out heat to the source plant.

rraustad commented on src/EnergyPlus/PlantLoopHeatPumpEIR.cc:
```diff
@@ -341,44 +559,60 @@ void EIRPlantLoopHeatPump::doPhysics(EnergyPlusData &state, Real64 currentLoad)
-                this->sizeSrcSideWSHP(state);
-            } else if (this->airSource) {
-                this->sizeSrcSideASHP(state);
-            }
```
sizing gets called multiple times when HVAC sizing simulation is performed

rraustad commented on src/EnergyPlus/PlantLoopHeatPumpEIR.cc:
```diff
@@ -402,7 +636,18 @@ void EIRPlantLoopHeatPump::onInitLoopEquip(EnergyPlusData &state, [[maybe_unused
             this->sourceSideDesignMassFlowRate = rho * this->sourceSideDesignVolFlowRate;
         }
 
-        this->envrnInit = false;
```
next sets up logicals to know how to handle VS pumps

rraustad commented on src/EnergyPlus/PlantLoopHeatPumpEIR.cc:
```diff
@@ -437,42 +682,129 @@ void EIRPlantLoopHeatPump::sizeLoadSide(EnergyPlusData &state)
-                                                             state.dataPlnt->PlantLoop(this->loadSidePlantLoc.loopNum).FluidName,
-                                                             loadSideInitTemp,
-                                                             state.dataPlnt->PlantLoop(this->loadSidePlantLoc.loopNum).FluidIndex,
-                                                             "EIRPlantLoopHeatPump::size()");
```
next calculate proper Cp and rho based on CW or HW intConvTemp

rraustad commented on src/EnergyPlus/PlantLoopHeatPumpEIR.cc:
```diff
@@ -437,42 +682,129 @@ void EIRPlantLoopHeatPump::sizeLoadSide(EnergyPlusData &state)
         // this first IF block is really just about calculating the local tmpCapacity and tmpLoadVolFlow values
         // these represent what the unit would size those to, whether it is doing auto-sizing or not
         if (state.dataSize->PlantSizData(pltLoadSizNum).DesVolFlowRate > DataHVACGlobals::SmallWaterVolFlow) {
             tmpLoadVolFlow = state.dataSize->PlantSizData(pltLoadSizNum).DesVolFlowRate * this->sizingFactor;
```
next is significant updates on sizing

rraustad commented on src/EnergyPlus/PlantLoopHeatPumpEIR.cc:
```diff
@@ -437,42 +682,129 @@ void EIRPlantLoopHeatPump::sizeLoadSide(EnergyPlusData &state)
         } else {
             if (this->referenceCapacityWasAutoSized) tmpCapacity = 0.0;
             if (this->loadSideDesignVolFlowRateWasAutoSized) tmpLoadVolFlow = 0.0;
         }
```
next is heat recovery HP connected to plant inlet/outlet used plant flow rate, and save temp capacity to use during subsequent sizing iterations

rraustad commented on src/EnergyPlus/PlantLoopHeatPumpEIR.cc:
```diff
@@ -696,13 +1030,22 @@ void EIRPlantLoopHeatPump::sizeSrcSideWSHP(EnergyPlusData &state)
             tmpSourceVolFlow = hardSizedSourceSideFlow;
         }
     }
```
next account for new sizing ratio input

rraustad commented on src/EnergyPlus/PlantLoopHeatPumpEIR.cc:
```diff
@@ -696,13 +1030,22 @@ void EIRPlantLoopHeatPump::sizeSrcSideWSHP(EnergyPlusData &state)
     // register the design volume flows with the plant, only doing half of source because the companion
     // is generally on the same loop
-    PlantUtilities::RegisterPlantCompDesignFlow(state, this->loadSideNodes.inlet, tmpLoadVolFlow);
-    PlantUtilities::RegisterPlantCompDesignFlow(state, this->sourceSideNodes.inlet, tmpSourceVolFlow / 0.5);
```
next, if not a heat recovery HP register plant flow rate, otherwise this HP does not add to plant sizing

rraustad commented on src/EnergyPlus/PlantLoopHeatPumpEIR.cc:
```diff
@@ -897,182 +1246,269 @@ void EIRPlantLoopHeatPump::processInputForEIRPLHP(EnergyPlusData &state)
-        for (auto instance = instancesValue.begin(); instance != instancesValue.end(); ++instance) {
-            auto const &fields = instance.value();
-            std::string const &thisObjectName = instance.key();
-            state.dataInputProcessing->inputProcessor->markObjectAsUsed(cCurrentModuleObject, thisObjectName);
```
next is input processing

rraustad commented on src/EnergyPlus/PlantLoopHeatPumpEIR.cc:
```diff
@@ -1106,6 +1542,32 @@ void EIRPlantLoopHeatPump::checkConcurrentOperation(EnergyPlusData &state)
     }
 }
 
```
next determines if valid heat recovery HP that is connected to a plant inlet or outlet

rraustad commented on src/EnergyPlus/PlantLoopHeatPumpEIR.cc:
```diff
@@ -1115,6 +1577,20 @@ void EIRPlantLoopHeatPump::oneTimeInit(EnergyPlusData &state)
         bool errFlag = false;
 
         // setup output variables
```
next set up new reports

rraustad commented on src/EnergyPlus/PlantLoopHeatPumpEIR.cc:
```diff
@@ -1209,6 +1685,41 @@ void EIRPlantLoopHeatPump::oneTimeInit(EnergyPlusData &state)
                                 "Heating",
                                 "Heat Pump",
                                 "Plant");
```
and defrost reports if defrost is applicable

rraustad commented on src/EnergyPlus/PlantLoopHeatPumpEIR.hh:
```diff
@@ -65,6 +65,36 @@ struct EnergyPlusData;
 
 namespace EIRPlantLoopHeatPumps {
 
```
set up new enums and stict variables

rraustad commented on src/EnergyPlus/PlantUtilities.cc:
```diff
@@ -1999,4 +2016,54 @@ bool verifyTwoNodeNumsOnSamePlantLoop(EnergyPlusData &state, int const nodeIndex
     return (matchedIndexA == matchedIndexB) && (matchedIndexA != 0); // only return true if both are equal and non-zero
 }
 
```
new worker function below to check VS pump configurations

rraustad commented on src/EnergyPlus/ReportCoilSelection.cc:
```diff
@@ -752,7 +752,7 @@ void ReportCoilSelection::doFinalProcessingOfCoilData(EnergyPlusData &state)
         }
 
         // call psych routine to flush out moist air metrics from those available
```
next just avoids reporting wierd data to table reports

rraustad commented on src/EnergyPlus/UnitarySystem.cc:
```diff
@@ -474,7 +474,7 @@ namespace UnitarySystems {
             }
         }
 
```
note as reminder that UnitarySystem can't set fan volume flow until components have sized. this is a defect that needs to be corrected.

rraustad commented on testfiles/CMakeLists.txt:
```diff
@@ -441,6 +441,8 @@ add_simulation_test(IDF_FILE PlantLoopChainHeating.idf EPW_FILE USA_IL_Chicago-O
 add_simulation_test(IDF_FILE PlantLoopHeatPump_EIR_AirSource.idf EPW_FILE USA_IL_Chicago-OHare.Intl.AP.725300_TMY3.epw)
 add_simulation_test(IDF_FILE PlantLoopHeatPump_EIR_WaterSource.idf EPW_FILE USA_IL_Chicago-OHare.Intl.AP.725300_TMY3.epw)
 add_simulation_test(IDF_FILE PlantLoopHeatPump_Fuel-Fired.idf EPW_FILE USA_IL_Chicago-OHare.Intl.AP.725300_TMY3.epw)
```
new example files and units tests are below

rraustad commented on testfiles/PlantLoopHeatPump_EIR_AirSource.idf:
```diff
@@ -313,8 +313,8 @@
 
   Curve:Quadratic,
     EIRCurveFuncPLR,         !- Name
```
EIR curves should be proportional to PLR

mjwitte replied:
I'm confused. Is this just improving the example file or are there accompanying code changes that change the basis? And then it would require a transition rule?

rraustad replied:
Yes, there was a code change to accompany this example file change. IIRC, the EIR curve wasn't being used. I'll check the code where this is now accounted for.

rraustad reviewed (commented):
walk-thru

rraustad commented:
Snapshot view of supervisory control from a new example file. Left top is building loads, right top is plant loads. Magnitude is similar between these plots, a good sign plant operation represents what is shown by building loads. Lower row starting at left shows dispatch of plant HPs (0=off, 1=heating, 2=cooling, 3=simultaneous cooling and heating. Then heat recovery HP (0 = off, 1 = heating led, 2 = cooling led). Then auxilliary boiler dispatch, minimal boiler operation is a good sign.

![image](https://github.com/NREL/EnergyPlus/assets/5778156/2c828efd-af45-4ecb-a5ae-5b3314be87db)

rraustad committed:
Merge branch 'develop' of https://github.com/NREL/EnergyPlus into NFP-Chiller-Plant-Support-for-Carbon-Initiatives

rraustad committed:
CppCheck EquipAndOperations

rraustad committed:
Remove unused function

rraustad committed:
CppCheck LoopSide

rraustad committed:
Clang format LoopSide

rraustad committed:
Update source and doc

rraustad committed:
Add to IO doc

mjwitte commented on idd/Energy+.idd.in:
```diff
@@ -74631,37 +74668,32 @@ HeatPump:PlantLoop:EIR:Heating,
    N1,  \field Load Side Reference Flow Rate
         \note This component is currently a constant-flow device, meaning it will always
         \note try to request the full design flow from the central plant manager.
-        \required-field
         \type real
         \minimum> 0.0
         \units m3/s
         \ip-units gal/min
         \autosizable
         \default autosize
    N2,  \field Source Side Reference Flow Rate
-        \required-field
         \type real
         \minimum> 0.0
         \units m3/s
         \ip-units gal/min
         \autosizable
         \default autosize
    N3,  \field Reference Capacity
-        \required-field
         \type real
         \minimum> 0.0
         \units W
         \autosizable
         \default autosize
    N4,  \field Reference Coefficient of Performance
-        \required-field
         \type real
         \minimum> 0.0
-        \default 7.5
         \units W/W
+        \default 3.0
    N5,  \field Sizing Factor
-        \note Multiplies the autosized capacity and flow rates
-        \required-field
+        \note Multiplies the autosized source side reference flow rate
```
Sizing factor doesn't multiply capacity now? I would expect the autosized capacity will also change.

rraustad replied:
Indirectly. The sizing factor multiplies the water flow rate. The capacity is calculated from the water flow rate. I could change that comment back but I think it's now clear what is getting multiplied.

mjwitte replied:
Yes, but it's not clear that the flow rate becomes the basis for the capacity sizing. I would suggest something like:
`\note Multiplies the autosized source side reference flow rate which is then used to autosize the capacity.`

mjwitte commented on idd/Energy+.idd.in:
```diff
@@ -83172,6 +83315,89 @@ PlantEquipmentOperation:OutdoorDewpointDifference,
+       \object-list ControlSchemeList
+  A7 , \field Dedicated Chilled Water Return Recovery Heat Pump Name
+       \type object-list
+       \object-list validPlantEquipmentNames
```
PLHPCoolingNames

mjwitte commented on idd/Energy+.idd.in:
```diff
@@ -83172,6 +83315,89 @@ PlantEquipmentOperation:OutdoorDewpointDifference,
+       \note enter name of HeatPump:PlantLoop:EIR:Cooling object to control chilled water return adding heat to hot water return
+  A8 , \field Dedicated Hot Water Return Recovery Heat Pump Name
+       \type object-list
+       \object-list validPlantEquipmentNames
```
PLHPHeatingNames

mjwitte reviewed (commented):
@rraustad Lots going on here: new fields and options in HeatPump:PlantLoop:EIR:Heating/Cooling plus new PlantEquipmentOperation:ChillerHeaterChangeover. 

How about this for the main PR title: "New plant heat pump features and supervisory controller"?

jcyuan2020 commented on doc/engineering-reference/src/building-system-simulation-system-manager/plant-condenser-loops.tex:
```diff
@@ -564,6 +564,64 @@ \subsubsection{Outdoor Drybulb Range Based Operation, Outdoor Wetbulb Range Base
+\dot Q_{BuildingHeating} = \dot Q_{ZonePredictedHeating} +  \dot Q_{AirSysVentHeating}  + \dot Q_{ProcessHeating}
+\end{equation}
+PlantEquipmentOperation:ChillerHeaterChangeover also determines current loads on the plant loops to use in control decisions.  Plant loading uses a SingleSetPoint scheme in much the same way as for \hyperref[component-setpoint-based-operation]{Component Setpoint Based Operation}.  Therefore single temperature setpoint values (Node().TempSetPoint) are used and applied as described in \hyperref[loo ...
+The supervisory control routine uses the results of building and plant loads at the current timestep to decide if the heat-pump-based hydronic plant should be operated in one of four modes: off, heating-only, cooling-only, or simultaneous-heating-and-cooling. If $\dot Q_{BuildingHeating}$ is less than zero and $\dot Q_{BuildingCooling}$ is also negative (cooling loads are negative) then based on  ...
```
One minor issue for the documentation here: in the TeX text file, the paragraphs need an actual blank line (or one extra line break, like Line 578 above) between paragraphs. Otherwise these 3-4 paragraphs here (from Line 609 to Line 613, for example) would all cramped into one long paragraph in the pdf file compiled.

rraustad committed:
First set of comments

rraustad committed:
Revert diff commit

rraustad commented:
After this last commit, there should be only 2 files with diffs. The PlantLoopHeatPumpEIR* files. These are the diffs from the water source file. The air source file has very small diffs.

[PlantLoopHeatPump_EIR_WaterSource.zip](https://github.com/NREL/EnergyPlus/files/12404133/PlantLoopHeatPump_EIR_WaterSource.zip)

The largest diff is the GHX heat transfer at 0.1%.

![image](https://github.com/NREL/EnergyPlus/assets/5778156/c89e1a4c-b1e9-431d-b430-ed0dcad2dc93)

rraustad committed:
Revert change to EIRfPLR definition

mjwitte reviewed (approved):
Doc changes look good (I didn't read them thoroughly but skimmed over them and they have all the right pieces and thorough explanations). Nice figures, too.

Unit tests run serially all pass.

The new example files have some significant warnings (max hvac iterations and plant loop temps out of range) that warrant some investigation post I/O-Freeze.

This adds an impressive new supervisory control. Assuming CI comes back clean, this looks good to go. @Myoldmopar do you want to review more?

Myoldmopar commented:
I don't.  I am thankful to @rraustad for reverting the curve change there, and look forward to revisiting that soon so we can get that resolved.  I checked everything out again and this looks great.  Merge at will if CI is happy.

mjwitte commented:
CI is all happy (except Mac which is AWOL) merging.

rraustad commented:
> I don't. I am thankful to @rraustad for reverting the curve change there, and look forward to revisiting that soon so we can get that resolved. I checked everything out again and this looks great. Merge at will if CI is happy.

I had an UnmetHours question on the EIR HP and looked at the code and was confused about the power equation. This comment made me remember why I thought I changed it but did not.

https://unmethours.com/question/98105/air-to-water-heat-pump-plr-performance-curve-issue/?answer=98123#post-id-98123

Added issue #10387.
